Index: Makefile
===================================================================
--- Makefile	(revisión: 3096)
+++ Makefile	(copia de trabajo)
@@ -1,7 +1,7 @@
 include build.mak
 include build/host-$(HOST_NAME).mak
 
-DIRS = pjlib pjlib-util pjnath third_party pjmedia pjsip pjsip-apps
+DIRS = pjlib pjlib-util pjnath third_party pjmedia pjsip
 
 ifdef MINSIZE
 MAKE_FLAGS := MINSIZE=1
Index: pjsip/build/Makefile
===================================================================
--- pjsip/build/Makefile	(revisión: 3096)
+++ pjsip/build/Makefile	(copia de trabajo)
@@ -102,7 +102,7 @@
 # Main entry
 #
 #
-TARGETS := pjsip pjsip-ua pjsip-simple pjsua-lib pjsip-test
+TARGETS := pjsip pjsip-ua pjsip-simple
 
 .PHONY: $(TARGETS)
 
Index: build.mak.in
===================================================================
--- build.mak.in	(revisión: 3096)
+++ build.mak.in	(copia de trabajo)
@@ -54,8 +54,7 @@
 	-L$(PJDIR)/pjsip/lib\
 	-L$(PJDIR)/third_party/lib\
 	@LDFLAGS@
-export APP_LDLIBS := -lpjsua-$(TARGET_NAME)\
-	-lpjsip-ua-$(TARGET_NAME)\
+export APP_LDLIBS := -lpjsip-ua-$(TARGET_NAME)\
 	-lpjsip-simple-$(TARGET_NAME)\
 	-lpjsip-$(TARGET_NAME)\
 	-lpjmedia-codec-$(TARGET_NAME)\
@@ -65,8 +64,7 @@
 	$(APP_THIRD_PARTY_LIBS)\
 	-lpj-$(TARGET_NAME)\
 	@LIBS@
-export APP_LIB_FILES = $(PJ_DIR)/pjsip/lib/libpjsua-$(LIB_SUFFIX) \
-	$(PJ_DIR)/pjsip/lib/libpjsip-ua-$(LIB_SUFFIX) \
+export APP_LIB_FILES = $(PJ_DIR)/pjsip/lib/libpjsip-ua-$(LIB_SUFFIX) \
 	$(PJ_DIR)/pjsip/lib/libpjsip-simple-$(LIB_SUFFIX) \
 	$(PJ_DIR)/pjsip/lib/libpjsip-$(LIB_SUFFIX) \
 	$(PJ_DIR)/pjmedia/lib/libpjmedia-codec-$(LIB_SUFFIX) \
