--- pjmedia/src/pjmedia/sdp_neg.c	2013-02-18 14:05:55.000000000 +0100
+++ pjmedia/src/pjmedia/sdp_neg.c	2013-02-18 14:07:13.000000000 +0100
@@ -1383,7 +1383,8 @@
     PJ_ASSERT_RETURN(neg, PJ_EINVAL);
 
     /* Must be in LOCAL_OFFER state. */
-    PJ_ASSERT_RETURN(neg->state == PJMEDIA_SDP_NEG_STATE_LOCAL_OFFER,
+    PJ_ASSERT_RETURN(neg->state == PJMEDIA_SDP_NEG_STATE_LOCAL_OFFER ||
+                     neg->state == PJMEDIA_SDP_NEG_STATE_REMOTE_OFFER,
 		     PJMEDIA_SDPNEG_EINSTATE);
 
     /* Reset state to done */
