--- pjsip/src/pjsip/sip_auth_msg.c	(revision 3148)
+++ pjsip/src/pjsip/sip_auth_msg.c	(revision 3149)
@@ -232,7 +232,8 @@
     char *endbuf = buf + size;
     const pjsip_parser_const_t *pc = pjsip_parser_const();
 
-    copy_advance_pair_quote_cond(buf, " realm=", 7, chal->realm, '"', '"');
+    /* Allow empty realm, see http://trac.pjsip.org/repos/ticket/1061 */
+    copy_advance_pair_quote(buf, " realm=", 7, chal->realm, '"', '"');
     copy_advance_pair_quote_cond(buf, ",domain=", 8, chal->domain, '"', '"');
     copy_advance_pair_quote_cond(buf, ",nonce=", 7, chal->nonce, '"', '"');
     copy_advance_pair_quote_cond(buf, ",opaque=", 8, chal->opaque, '"', '"');
--- pjsip/src/pjsip/sip_auth_client.c	(revision 3148)
+++ pjsip/src/pjsip/sip_auth_client.c	(revision 3149)
@@ -393,7 +393,9 @@
 	cached_auth->nc = 1;
 
 	/* Save realm. */
+	/* Note: allow empty realm (http://trac.pjsip.org/repos/ticket/1061)
 	pj_assert(cached_auth->realm.slen != 0);
+	*/
 	if (cached_auth->realm.slen == 0) {
 	    pj_strdup(ses_pool, &cached_auth->realm, 
 		      &hdr->challenge.digest.realm);
