===================================================================
--- pjsip/src/pjsip/sip_dialog.c	(revision 2825)
+++ pjsip/src/pjsip/sip_dialog.c	(working copy)
@@ -1733,8 +1733,7 @@
 	(dlg->role==PJSIP_ROLE_UAC &&
 	 !dlg->uac_has_2xx &&
 	 res_code/100 <= 2 &&
-	 pjsip_method_creates_dialog(&rdata->msg_info.cseq->method) &&
-	 pj_strcmp(&dlg->remote.info->tag, &rdata->msg_info.to->tag)))
+	 pjsip_method_creates_dialog(&rdata->msg_info.cseq->method)))
     {
 	pjsip_contact_hdr *contact;
 
