===================================================================
--- pjsip/src/pjsip-ua/sip_inv.c	(revision 2672)
+++ pjsip/src/pjsip-ua/sip_inv.c	(working copy)
@@ -3523,18 +3523,10 @@
     {
 
 	/*
-	 * Handle strandled incoming CANCEL.
+	 * Handle strandled incoming CANCEL or CANCEL for re-INVITE
 	 */
-	pjsip_rx_data *rdata = e->body.tsx_state.src.rdata;
-	pjsip_tx_data *tdata;
-	pj_status_t status;
+   inv_respond_incoming_cancel(inv, tsx, e->body.tsx_state.src.rdata);
 
-	status = pjsip_dlg_create_response(dlg, rdata, 200, NULL, &tdata);
-	if (status != PJ_SUCCESS) return;
-
-	status = pjsip_dlg_send_response(dlg, tsx, tdata);
-	if (status != PJ_SUCCESS) return;
-
     }
     else if (tsx->method.id == PJSIP_INVITE_METHOD &&
 	     tsx->role == PJSIP_ROLE_UAS)
