--- aconfigure	2008-09-27 15:16:35.000000000 +0200
+++ aconfigure	2011-09-13 11:57:10.000000000 +0200
@@ -308,7 +308,7 @@
 # include <unistd.h>
 #endif"
 
-ac_subst_vars='SHELL PATH_SEPARATOR PACKAGE_NAME PACKAGE_TARNAME PACKAGE_VERSION PACKAGE_STRING PACKAGE_BUGREPORT exec_prefix prefix program_transform_name bindir sbindir libexecdir datadir sysconfdir sharedstatedir localstatedir libdir includedir oldincludedir infodir mandir build_alias host_alias target_alias DEFS ECHO_C ECHO_N ECHO_T LIBS build build_cpu build_vendor build_os host host_cpu host_vendor host_os target target_cpu target_vendor target_os CC CFLAGS LDFLAGS CPPFLAGS ac_ct_CC EXEEXT OBJEXT CXX CXXFLAGS ac_ct_CXX ac_pjdir ac_build_mak_vars CPP EGREP ac_os_objs ac_pjmedia_snd ac_pa_cflags ac_pa_use_alsa ac_pa_use_oss ac_no_small_filter ac_no_large_filter ac_no_speex_aec ac_no_g711_codec ac_no_l16_codec ac_no_gsm_codec ac_no_g722_codec ac_no_speex_codec ac_no_ilbc_codec ac_no_ssl openssl_h_present libssl_present libcrypto_present ac_cross_compile ac_linux_poll ac_host ac_main_obj LIBOBJS LTLIBOBJS'
+ac_subst_vars='SHELL PATH_SEPARATOR PACKAGE_NAME PACKAGE_TARNAME PACKAGE_VERSION PACKAGE_STRING PACKAGE_BUGREPORT exec_prefix prefix program_transform_name bindir sbindir libexecdir datadir sysconfdir sharedstatedir localstatedir libdir includedir oldincludedir infodir mandir build_alias host_alias target_alias DEFS ECHO_C ECHO_N ECHO_T LIBS build build_cpu build_vendor build_os host host_cpu host_vendor host_os target target_cpu target_vendor target_os CC CFLAGS LDFLAGS CPPFLAGS ac_ct_CC EXEEXT OBJEXT CXX CXXFLAGS ac_ct_CXX ac_pjdir ac_build_mak_vars CPP EGREP ac_os_objs ac_pjmedia_snd ac_pa_cflags ac_pa_use_alsa ac_pa_use_oss ac_no_small_filter ac_no_large_filter ac_no_speex_aec ac_no_g711_codec ac_no_l16_codec ac_no_gsm_codec ac_no_g722_codec ac_no_speex_codec ac_no_ilbc_codec ac_webrtc_platform ac_no_ssl openssl_h_present libssl_present libcrypto_present ac_cross_compile ac_linux_poll ac_host ac_main_obj LIBOBJS LTLIBOBJS'
 ac_subst_files=''
 
 # Initialize some variables set by options.
@@ -1452,7 +1452,7 @@
   program_prefix=${target_alias}-
                                                   ac_config_headers="$ac_config_headers pjlib/include/pj/compat/os_auto.h pjlib/include/pj/compat/m_auto.h pjmedia/include/pjmedia/config_auto.h pjmedia/include/pjmedia-codec/config_auto.h pjsip/include/pjsip/sip_autoconf.h"
 
-                                                                                ac_config_files="$ac_config_files build.mak build/os-auto.mak pjlib/build/os-auto.mak pjlib-util/build/os-auto.mak pjmedia/build/os-auto.mak pjsip/build/os-auto.mak third_party/build/portaudio/os-auto.mak third_party/build/os-auto.mak"
+                                                                                          ac_config_files="$ac_config_files build.mak build/os-auto.mak pjlib/build/os-auto.mak pjlib-util/build/os-auto.mak pjmedia/build/os-auto.mak pjsip/build/os-auto.mak third_party/build/portaudio/os-auto.mak third_party/build/webrtc/os-auto.mak third_party/build/os-auto.mak"
 
 
 
@@ -10963,6 +10963,24 @@
 
 
 
+case $target in
+ *darwin*)
+       ac_webrtc_platform=webrtc_darwinos
+       ;;
+ *mingw*)
+       ac_webrtc_platform=webrtc_win32
+       ;;
+ *linux*)
+       ac_webrtc_platform=webrtc_linux
+       ;;
+ *)
+       ac_webrtc_platform=null
+       ;;
+esac
+
+
+
+
 # Check whether --enable-ssl or --disable-ssl was given.
 if test "${enable_ssl+set}" = set; then
   enableval="$enable_ssl"
@@ -11921,6 +11939,7 @@
   "pjmedia/build/os-auto.mak" ) CONFIG_FILES="$CONFIG_FILES pjmedia/build/os-auto.mak" ;;
   "pjsip/build/os-auto.mak" ) CONFIG_FILES="$CONFIG_FILES pjsip/build/os-auto.mak" ;;
   "third_party/build/portaudio/os-auto.mak" ) CONFIG_FILES="$CONFIG_FILES third_party/build/portaudio/os-auto.mak" ;;
+  "third_party/build/webrtc/os-auto.mak" ) CONFIG_FILES="$CONFIG_FILES third_party/build/webrtc/os-auto.mak" ;;
   "third_party/build/os-auto.mak" ) CONFIG_FILES="$CONFIG_FILES third_party/build/os-auto.mak" ;;
   "pjlib/include/pj/compat/os_auto.h" ) CONFIG_HEADERS="$CONFIG_HEADERS pjlib/include/pj/compat/os_auto.h" ;;
   "pjlib/include/pj/compat/m_auto.h" ) CONFIG_HEADERS="$CONFIG_HEADERS pjlib/include/pj/compat/m_auto.h" ;;
@@ -12051,6 +12070,7 @@
 s,@ac_no_g722_codec@,$ac_no_g722_codec,;t t
 s,@ac_no_speex_codec@,$ac_no_speex_codec,;t t
 s,@ac_no_ilbc_codec@,$ac_no_ilbc_codec,;t t
+s,@ac_webrtc_platform@,$ac_webrtc_platform,;t t
 s,@ac_no_ssl@,$ac_no_ssl,;t t
 s,@openssl_h_present@,$openssl_h_present,;t t
 s,@libssl_present@,$libssl_present,;t t
--- aconfigure.ac	2008-09-27 15:16:35.000000000 +0200
+++ aconfigure.ac	2011-09-13 11:51:13.000000000 +0200
@@ -18,6 +18,7 @@
 		 pjmedia/build/os-auto.mak
 		 pjsip/build/os-auto.mak
 		 third_party/build/portaudio/os-auto.mak 
+		 third_party/build/webrtc/os-auto.mak 
 		 third_party/build/os-auto.mak
 	        ])
 
@@ -719,6 +720,28 @@
 
 
 dnl ##########################################
+dnl #
+dnl # WebRTC AEC Support
+dnl #
+
+AC_SUBST(ac_webrtc_platform)
+case $target in
+ *darwin*)
+       ac_webrtc_platform=webrtc_darwinos
+       ;;
+ *mingw*)
+       ac_webrtc_platform=webrtc_win32
+       ;;
+ *linux*)
+       ac_webrtc_platform=webrtc_linux
+       ;;
+ *)
+       ac_webrtc_platform=null
+       ;;
+esac
+
+
+dnl ##########################################
 dnl #
 dnl # PJSIP CONFIG
 dnl #
--- build/rules.mak	2008-07-29 13:42:20.000000000 +0200
+++ build/rules.mak	2011-09-13 09:58:04.000000000 +0200
@@ -127,6 +127,11 @@
 		$(CC_OUT)$(subst /,$(HOST_PSEP),$@) \
 		$(subst /,$(HOST_PSEP),$<)
 
+$(OBJDIR)/%$(OBJEXT): $(SRCDIR)/%.cc
+	$(CC) $($(APP)_CXXFLAGS) \
+		$(CC_OUT)$(subst /,$(HOST_PSEP),$@) \
+		$(subst /,$(HOST_PSEP),$<)
+
 $(OBJDIRS):
 	$(subst @@,$(subst /,$(HOST_PSEP),$@),$(HOST_MKDIR)) 
 
--- build.mak.in	2011-08-22 17:24:20.000000000 +0200
+++ build.mak.in	2011-09-13 09:58:04.000000000 +0200
@@ -13,6 +13,10 @@
 export APP_THIRD_PARTY_LIBS := -lresample-$(TARGET_NAME) -lmilenage-$(TARGET_NAME) -lsrtp-$(TARGET_NAME)
 export APP_THIRD_PARTY_LIB_FILES = $(PJ_DIR)/third_party/lib/libresample-$(LIB_SUFFIX) $(PJ_DIR)/third_party/lib/libmilenage-$(LIB_SUFFIX) $(PJ_DIR)/third_party/lib/libsrtp-$(LIB_SUFFIX)
 
+# WebRTC
+APP_THIRD_PARTY_LIBS += -lwebrtcaec-$(TARGET_NAME)
+APP_THIRD_PARTY_LIB_FILES += $(PJ_DIR)/third_party/lib/libwebrtcaec-$(LIB_SUFFIX)
+
 ifneq (@ac_no_gsm_codec@,1)
 APP_THIRD_PARTY_LIBS += -lgsmcodec-$(TARGET_NAME)
 APP_THIRD_PARTY_LIB_FILES += $(PJ_DIR)/third_party/lib/libgsmcodec-$(LIB_SUFFIX)
--- pjmedia/build/os-auto.mak.in	2008-07-29 01:40:58.000000000 +0200
+++ pjmedia/build/os-auto.mak.in	2011-09-13 15:12:04.000000000 +0200
@@ -61,6 +61,9 @@
 
 endif
 
+export CFLAGS += -I$(THIRD_PARTY)/webrtc/src
+export PJMEDIA_OBJS += echo_webrtc_aec.o
+
 ifeq ($(AC_NO_ILBC_CODEC),1)
 export CFLAGS += -DPJMEDIA_HAS_ILBC_CODEC=0
 else
--- pjmedia/build/os-darwinos.mak	2008-03-17 15:24:21.000000000 +0100
+++ pjmedia/build/os-darwinos.mak	2011-09-13 10:24:06.000000000 +0200
@@ -58,6 +58,9 @@
 
 endif
 
+export CFLAGS += -I$(THIRD_PARTY)/webrtc/src
+export PJMEDIA_OBJS += echo_webrtc_aec.o
+
 ifeq ($(AC_NO_ILBC_CODEC),1)
 export CFLAGS += -DPJMEDIA_HAS_ILBC_CODEC=0
 else
--- pjmedia/build/os-linux.mak	2008-03-17 15:24:21.000000000 +0100
+++ pjmedia/build/os-linux.mak	2011-09-13 10:24:17.000000000 +0200
@@ -52,6 +52,9 @@
 
 endif
 
+export CFLAGS += -I$(THIRD_PARTY)/webrtc/src
+export PJMEDIA_OBJS += echo_webrtc_aec.o
+
 ifeq ($(AC_NO_ILBC_CODEC),1)
 export CFLAGS += -DPJMEDIA_HAS_ILBC_CODEC=0
 else
--- pjmedia/build/os-win32.mak	2008-03-17 15:24:21.000000000 +0100
+++ pjmedia/build/os-win32.mak	2011-09-13 15:11:56.000000000 +0200
@@ -54,6 +54,9 @@
 
 endif
 
+export CFLAGS += -I$(THIRD_PARTY)/webrtc/src
+export PJMEDIA_OBJS += echo_webrtc_aec.o
+
 ifeq ($(AC_NO_ILBC_CODEC),1)
 export CFLAGS += -DPJMEDIA_HAS_ILBC_CODEC=0
 else
--- pjmedia/include/pjmedia/echo.h	2008-12-23 18:27:53.000000000 +0100
+++ pjmedia/include/pjmedia/echo.h	2011-09-13 09:58:03.000000000 +0200
@@ -79,6 +79,12 @@
     PJMEDIA_ECHO_SIMPLE	= 2,
 
     /**
+     * Force to use WebRTC AEC as the backend echo canceller algorithm.
+     * This setting is mutually exclusive with PJMEDIA_ECHO_SIMPLE & PJMEDIA_ECHO_SPEEX.
+     */
+    PJMEDIA_ECHO_WEBRTC = 3,
+
+    /**
      * For internal use.
      */
     PJMEDIA_ECHO_ALGO_MASK = 15,
--- pjmedia/src/pjmedia/echo_common.c	2011-08-22 17:24:20.000000000 +0200
+++ pjmedia/src/pjmedia/echo_common.c	2011-09-13 09:58:04.000000000 +0200
@@ -115,6 +115,22 @@
 };
 #endif
 
+
+/*
+ * WebRTC prototypes
+ */
+#if defined(PJMEDIA_HAS_WEBRTC_AEC) && PJMEDIA_HAS_WEBRTC_AEC!=0
+static struct ec_operations webrtc_aec_op =
+{
+    "WEBRTC AEC",
+    &webrtc_aec_create,
+    &webrtc_aec_destroy,
+    &webrtc_aec_reset,
+    &webrtc_aec_cancel_echo
+};
+#endif
+
+
 /*
  * Create the echo canceller. 
  */
@@ -165,6 +181,13 @@
 	ec->op = &speex_aec_op;
 #endif
 
+#if defined(PJMEDIA_HAS_WEBRTC_AEC) && PJMEDIA_HAS_WEBRTC_AEC!=0
+    } else if ((options & PJMEDIA_ECHO_ALGO_MASK) == PJMEDIA_ECHO_WEBRTC ||
+	       (options & PJMEDIA_ECHO_ALGO_MASK) == PJMEDIA_ECHO_DEFAULT) 
+    {
+	ec->op = &webrtc_aec_op;
+#endif
+
 #if defined(PJMEDIA_HAS_INTEL_IPP_AEC) && PJMEDIA_HAS_INTEL_IPP_AEC!=0
     } else if ((options & PJMEDIA_ECHO_ALGO_MASK) == PJMEDIA_ECHO_IPP ||
 	       (options & PJMEDIA_ECHO_ALGO_MASK) == PJMEDIA_ECHO_DEFAULT)
--- pjmedia/src/pjmedia/echo_internal.h	2008-12-23 18:27:53.000000000 +0100
+++ pjmedia/src/pjmedia/echo_internal.h	2011-09-13 09:58:04.000000000 +0200
@@ -72,6 +72,20 @@
 					 unsigned options,
 					 void *reserved );
 
+PJ_DECL(pj_status_t) webrtc_aec_create(pj_pool_t *pool,
+			               unsigned clock_rate,
+			               unsigned channel_count,
+			               unsigned samples_per_frame,
+			               unsigned tail_ms,
+			               unsigned options,
+			               void **p_echo );
+PJ_DECL(pj_status_t) webrtc_aec_destroy(void *state );
+PJ_DECL(void) webrtc_aec_reset(void *state );
+PJ_DECL(pj_status_t) webrtc_aec_cancel_echo(void *state,
+					    pj_int16_t *rec_frm,
+					    const pj_int16_t *play_frm,
+					    unsigned options,
+					    void *reserved );
 
 PJ_END_DECL
 
--- pjmedia/src/pjmedia/echo_webrtc_aec.c	1970-01-01 01:00:00.000000000 +0100
+++ pjmedia/src/pjmedia/echo_webrtc_aec.c	2011-09-15 09:59:11.000000000 +0200
@@ -0,0 +1,263 @@
+/**
+ * Copyright (C) 2010 Regis Montoya (aka r3gis - www.r3gis.fr)
+ * This file is part of pjsip_android.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include <pjmedia/echo.h>
+#include <pjmedia/errno.h>
+#include <pj/assert.h>
+#include <pj/log.h>
+#include <pj/pool.h>
+
+
+#if defined(PJMEDIA_HAS_WEBRTC_AEC) && PJMEDIA_HAS_WEBRTC_AEC != 0
+
+#ifndef PJMEDIA_WEBRTC_AEC_AGGRESSIVENESS
+    #define PJMEDIA_WEBRTC_AEC_AGGRESSIVENESS 1
+#endif
+
+#ifndef PJMEDIA_WEBRTC_NS_POLICY
+    #define PJMEDIA_WEBRTC_NS_POLICY 0
+#endif
+
+#define WEBRTC_SUBFRAME_LEN   160    // WebRTC AEC only allows max 160 samples/frame
+
+#define THIS_FILE    "echo_webrtc_aec.c"
+
+#include <third_party/webrtc/src/modules/audio_processing/aec/main/interface/echo_cancellation.h>
+#include <third_party/webrtc/src/modules/audio_processing/ns/main/interface/noise_suppression.h>
+
+#include "echo_internal.h"
+
+typedef struct webrtc_ec
+{
+    void        *AEC_inst;
+    void        *NS_inst;
+    unsigned    samples_per_frame;
+    unsigned	echo_tail;
+    unsigned	echo_skew;
+    unsigned    clock_rate;
+    pj_int16_t	*tmp_frame;
+    pj_int16_t	*tmp_frame2;
+} webrtc_ec;
+
+
+
+static void print_webrtc_aec_error(const char* tag, void *AEC_inst) {
+    unsigned status = WebRtcAec_get_error_code(AEC_inst);
+    PJ_LOG(4, (THIS_FILE, "WebRTC AEC ERROR (%s) %d ", tag, status));
+}
+
+
+/*
+ * Create the AEC.
+ */
+PJ_DEF(pj_status_t) webrtc_aec_create(pj_pool_t *pool,
+				      unsigned clock_rate,
+				      unsigned channel_count,
+				      unsigned samples_per_frame,
+				      unsigned tail_ms,
+				      unsigned options,
+				      void **p_echo )
+{
+    webrtc_ec *echo;
+    int status;
+
+    *p_echo = NULL;
+
+    if (clock_rate != 16000) {
+        PJ_LOG(4, (THIS_FILE, "Unsupported sample rate: %d", clock_rate));
+        return PJ_EINVAL;
+    }
+
+    echo = PJ_POOL_ZALLOC_T(pool, webrtc_ec);
+    PJ_ASSERT_RETURN(echo != NULL, PJ_ENOMEM);
+
+    // Alloc memory
+    status = WebRtcAec_Create(&echo->AEC_inst);
+    if(status != 0) {
+    	return PJ_ENOMEM;
+    }
+
+    // Init
+    status = WebRtcAec_Init(echo->AEC_inst, clock_rate, clock_rate);
+    if(status != 0) {
+        if (echo->AEC_inst) {
+            print_webrtc_aec_error("Init", echo->AEC_inst);
+        }
+    	WebRtcAec_Free(echo->AEC_inst);
+    	return PJ_EBUG;
+    }
+
+    AecConfig aec_config;
+    aec_config.nlpMode = PJMEDIA_WEBRTC_AEC_AGGRESSIVENESS;
+    aec_config.skewMode = kAecTrue;
+    aec_config.metricsMode = kAecFalse;
+
+    status = WebRtcAec_set_config(echo->AEC_inst, aec_config);
+    if(status != 0) {
+        print_webrtc_aec_error("Init config", echo->AEC_inst);
+        WebRtcAec_Free(echo->AEC_inst);
+    	return PJ_EBUG;
+    }
+
+    status = WebRtcNs_Create(&echo->NS_inst);
+    if(status != 0) {
+    	return PJ_ENOMEM;
+    }
+
+    status = WebRtcNs_Init(echo->NS_inst, clock_rate);
+    if(status != 0) {
+        if (echo->NS_inst) {
+    	    WebRtcNs_Free(echo->NS_inst);
+            PJ_LOG(4, (THIS_FILE, "Could not initialize noise suppressor"));
+        }
+    	return PJ_EBUG;
+    }
+
+    status = WebRtcNs_set_policy(echo->NS_inst, PJMEDIA_WEBRTC_NS_POLICY);
+    if (status != 0) {
+        PJ_LOG(4, (THIS_FILE, "Could not set noise suppressor policy"));
+    }
+
+    echo->samples_per_frame = samples_per_frame;
+    echo->echo_tail = tail_ms;
+    echo->echo_skew = 0;
+    echo->clock_rate = clock_rate;
+
+    /* Create temporary frames for echo cancellation */
+    echo->tmp_frame = (pj_int16_t*) pj_pool_zalloc(pool, 2*samples_per_frame);
+    PJ_ASSERT_RETURN(echo->tmp_frame != NULL, PJ_ENOMEM);
+    echo->tmp_frame2 = (pj_int16_t*) pj_pool_zalloc(pool, 2*samples_per_frame);
+    PJ_ASSERT_RETURN(echo->tmp_frame2 != NULL, PJ_ENOMEM);
+
+    /* Done */
+    *p_echo = echo;
+    return PJ_SUCCESS;
+
+}
+
+
+/*
+ * Destroy AEC
+ */
+PJ_DEF(pj_status_t) webrtc_aec_destroy(void *state )
+{
+    webrtc_ec *echo = (webrtc_ec*) state;
+    PJ_ASSERT_RETURN(echo && echo->AEC_inst && echo->NS_inst, PJ_EINVAL);
+
+    if (echo->AEC_inst) {
+    	WebRtcAec_Free(echo->AEC_inst);
+    	echo->AEC_inst = NULL;
+    }
+    if (echo->NS_inst) {
+        WebRtcNs_Free(echo->NS_inst);
+        echo->NS_inst = NULL;
+    }
+
+    return PJ_SUCCESS;
+}
+
+
+/*
+ * Reset AEC
+ */
+PJ_DEF(void) webrtc_aec_reset(void *state )
+{
+    webrtc_ec *echo = (webrtc_ec*) state;
+    pj_assert(echo != NULL);
+    int status;
+    /* re-initialize the EC */
+    status = WebRtcAec_Init(echo->AEC_inst, echo->clock_rate, echo->clock_rate);
+    if(status != 0) {
+        print_webrtc_aec_error("re-Init", echo->AEC_inst);
+        return;
+    } else {
+        AecConfig aec_config;
+        aec_config.nlpMode = PJMEDIA_WEBRTC_AEC_AGGRESSIVENESS;
+        aec_config.skewMode = kAecTrue;
+        aec_config.metricsMode = kAecFalse;
+
+        status = WebRtcAec_set_config(echo->AEC_inst, aec_config);
+        if(status != 0) {
+            print_webrtc_aec_error("re-Init config", echo->AEC_inst);
+            return;
+        }
+    }
+    PJ_LOG(4, (THIS_FILE, "WebRTC AEC reset succeeded"));
+}
+
+
+/*
+ * Perform echo cancellation.
+ */
+PJ_DEF(pj_status_t) webrtc_aec_cancel_echo( void *state,
+					    pj_int16_t *rec_frm,
+					    const pj_int16_t *play_frm,
+					    unsigned options,
+					    void *reserved )
+{
+    webrtc_ec *echo = (webrtc_ec*) state;
+    int i;
+    int status;
+
+    /* Sanity checks */
+    PJ_ASSERT_RETURN(echo && rec_frm && play_frm && options==0 && reserved==NULL, PJ_EINVAL);
+
+    for(i=0; i < echo->samples_per_frame; i+= WEBRTC_SUBFRAME_LEN) {
+        /* Noise suppression */
+        status = WebRtcNs_Process(echo->NS_inst,
+                                  (WebRtc_Word16 *) (&rec_frm[i]),
+                                  NULL,
+                                  (WebRtc_Word16 *) (&echo->tmp_frame[i]),
+                                  NULL);
+        if (status != 0) {
+            PJ_LOG(4, (THIS_FILE, "Error suppressing noise"));
+            return PJ_EBUG;
+        }
+
+        /* Feed farend buffer */
+	status = WebRtcAec_BufferFarend(echo->AEC_inst, &play_frm[i], WEBRTC_SUBFRAME_LEN);
+	if(status != 0) {
+	    print_webrtc_aec_error("buffer farend", echo->AEC_inst);
+	    return PJ_EBUG;
+	}
+
+	/* Process echo cancellation */
+        status = WebRtcAec_Process(echo->AEC_inst,
+                                   (WebRtc_Word16 *) (&echo->tmp_frame[i]),
+                                   NULL,
+                                   (WebRtc_Word16 *) (&echo->tmp_frame2[i]),
+                                   NULL,
+                                   WEBRTC_SUBFRAME_LEN,
+                                   echo->echo_tail,
+                                   echo->echo_skew);
+        if(status != 0) {
+            print_webrtc_aec_error("process echo", echo->AEC_inst);
+	    return PJ_EBUG;
+        }
+    }
+
+
+    /* Copy temporary buffer back to original rec_frm */
+    pjmedia_copy_samples(rec_frm, echo->tmp_frame2, echo->samples_per_frame);
+
+    return PJ_SUCCESS;
+
+}
+
+
+#endif
--- third_party/build/os-auto.mak.in	2007-05-15 12:42:56.000000000 +0200
+++ third_party/build/os-auto.mak.in	2011-09-13 10:23:43.000000000 +0200
@@ -14,3 +14,6 @@
 ifneq ($(findstring pa,@ac_pjmedia_snd@),)
 DIRS += portaudio
 endif
+
+DIRS += webrtc
+
--- third_party/build/os-darwinos.mak	2007-10-10 13:37:56.000000000 +0200
+++ third_party/build/os-darwinos.mak	2011-09-13 10:23:59.000000000 +0200
@@ -2,5 +2,6 @@
 DIRS += ilbc
 DIRS += speex
 DIRS += portaudio
+DIRS += webrtc
 
 
--- third_party/build/os-linux.mak	2007-10-10 13:37:56.000000000 +0200
+++ third_party/build/os-linux.mak	2011-09-13 10:23:54.000000000 +0200
@@ -2,5 +2,6 @@
 DIRS += ilbc
 DIRS += speex
 DIRS += portaudio
+DIRS += webrtc
 
 
--- third_party/build/webrtc/Makefile	1970-01-01 01:00:00.000000000 +0100
+++ third_party/build/webrtc/Makefile	2011-09-14 09:32:12.000000000 +0200
@@ -0,0 +1,159 @@
+include ../../../build.mak
+include ../../../build/common.mak
+
+export LIBDIR := ../../lib
+
+RULES_MAK := $(PJDIR)/build/rules.mak
+
+export WEBRTCAEC_LIB := ../../lib/libwebrtcaec-$(TARGET_NAME)$(LIBEXT)
+
+WEBRTC_DEFS := \
+	-D__STDC_FORMAT_MACROS \
+	-DDYNAMIC_ANNOTATIONS_ENABLED=1 \
+	-D_DEBUG
+
+WEBRTC_CFLAGS := \
+	-fno-exceptions \
+	-Wall \
+	-Wno-unused-parameter \
+	-Wno-missing-field-initializers \
+	-D_FILE_OFFSET_BITS=64 \
+	-fvisibility=hidden \
+	-pipe \
+	-fno-strict-aliasing
+
+WEBRTC_CFLAGS_CC := \
+	-fno-threadsafe-statics \
+	-fvisibility-inlines-hidden \
+	-Wsign-compare
+
+WEBRTC_SRC_BASE := ../../webrtc
+
+WEBRTCAEC_INCS := \
+		   $(CC_INC)$(WEBRTC_SRC_BASE)/src \
+		   $(CC_INC)$(WEBRTC_SRC_BASE)/src/common_audio/signal_processing_library/main/interface \
+		   $(CC_INC)$(WEBRTC_SRC_BASE)/src/modules/audio_processing/aec/main/interface \
+		   $(CC_INC)$(WEBRTC_SRC_BASE)/src/modules/audio_processing/ns/main/interface \
+		   $(CC_INC)$(WEBRTC_SRC_BASE)/src/modules/audio_processing/utility \
+		   $(CC_INC)$(WEBRTC_SRC_BASE)/src/system_wrappers/source/spreadsortlib \
+		   $(CC_INC)$(WEBRTC_SRC_BASE)/src/system_wrappers/interface
+
+export _CFLAGS 	:= $(CC_CFLAGS) $(OS_CFLAGS) $(HOST_CFLAGS) $(M_CFLAGS) $(CFLAGS) \
+		   $(WEBRTC_DEFS) $(WEBRTC_CFLAGS) \
+		   $(CC_INC). \
+		   $(CC_INC)../../../pjlib/include \
+		   $(WEBRTCAEC_INCS)
+export _CXXFLAGS:= $(_CFLAGS) $(CC_CXXFLAGS) $(OS_CXXFLAGS) $(M_CXXFLAGS) $(HOST_CXXFLAGS) $(CXXFLAGS) \
+		   $(WEBRTC_DEFS) $(WEBRTC_CFLAGS_CC)
+export _LDFLAGS := $(CC_LDFLAGS) $(OS_LDFLAGS) $(M_LDFLAGS) $(HOST_LDFLAGS) \
+		   $(LDFLAGS)
+
+
+AEC_OBJS := modules/audio_processing/aec/main/source/echo_cancellation.o \
+	    modules/audio_processing/aec/main/source/aec_core.o \
+	    modules/audio_processing/aec/main/source/aec_core_sse2.o \
+	    modules/audio_processing/aec/main/source/aec_rdft.o \
+	    modules/audio_processing/aec/main/source/aec_rdft_sse2.o \
+	    modules/audio_processing/aec/main/source/resampler.o
+
+APM_OBJS := modules/audio_processing/utility/ring_buffer.o \
+	    modules/audio_processing/utility/fft4g.o
+
+SPL_OBJS := common_audio/signal_processing_library/main/source/auto_corr_to_refl_coef.o \
+	    common_audio/signal_processing_library/main/source/auto_correlation.o \
+	    common_audio/signal_processing_library/main/source/complex_fft.o \
+	    common_audio/signal_processing_library/main/source/complex_ifft.o \
+	    common_audio/signal_processing_library/main/source/complex_bit_reverse.o \
+	    common_audio/signal_processing_library/main/source/copy_set_operations.o \
+	    common_audio/signal_processing_library/main/source/cos_table.o \
+	    common_audio/signal_processing_library/main/source/cross_correlation.o \
+	    common_audio/signal_processing_library/main/source/division_operations.o \
+	    common_audio/signal_processing_library/main/source/dot_product_with_scale.o \
+	    common_audio/signal_processing_library/main/source/downsample_fast.o \
+	    common_audio/signal_processing_library/main/source/energy.o \
+	    common_audio/signal_processing_library/main/source/filter_ar.o \
+	    common_audio/signal_processing_library/main/source/filter_ar_fast_q12.o \
+	    common_audio/signal_processing_library/main/source/filter_ma_fast_q12.o \
+	    common_audio/signal_processing_library/main/source/get_hanning_window.o \
+	    common_audio/signal_processing_library/main/source/get_scaling_square.o \
+	    common_audio/signal_processing_library/main/source/hanning_table.o \
+	    common_audio/signal_processing_library/main/source/ilbc_specific_functions.o \
+	    common_audio/signal_processing_library/main/source/levinson_durbin.o \
+	    common_audio/signal_processing_library/main/source/lpc_to_refl_coef.o \
+	    common_audio/signal_processing_library/main/source/min_max_operations.o \
+	    common_audio/signal_processing_library/main/source/randn_table.o \
+	    common_audio/signal_processing_library/main/source/randomization_functions.o \
+	    common_audio/signal_processing_library/main/source/refl_coef_to_lpc.o \
+	    common_audio/signal_processing_library/main/source/resample.o \
+	    common_audio/signal_processing_library/main/source/resample_48khz.o \
+	    common_audio/signal_processing_library/main/source/resample_by_2.o \
+	    common_audio/signal_processing_library/main/source/resample_by_2_internal.o \
+	    common_audio/signal_processing_library/main/source/resample_fractional.o \
+	    common_audio/signal_processing_library/main/source/sin_table.o \
+	    common_audio/signal_processing_library/main/source/sin_table_1024.o \
+	    common_audio/signal_processing_library/main/source/spl_sqrt.o \
+	    common_audio/signal_processing_library/main/source/spl_sqrt_floor.o \
+	    common_audio/signal_processing_library/main/source/spl_version.o \
+	    common_audio/signal_processing_library/main/source/splitting_filter.o \
+	    common_audio/signal_processing_library/main/source/sqrt_of_one_minus_x_squared.o \
+	    common_audio/signal_processing_library/main/source/vector_scaling_operations.o \
+	    common_audio/signal_processing_library/main/source/webrtc_fft_t_1024_8.o \
+	    common_audio/signal_processing_library/main/source/webrtc_fft_t_rad.o
+
+SWR_OBJS := system_wrappers/source/aligned_malloc.o \
+            system_wrappers/source/atomic32.o \
+            system_wrappers/source/condition_variable.o \
+            system_wrappers/source/cpu.o \
+            system_wrappers/source/cpu_features.o \
+            system_wrappers/source/critical_section.o \
+            system_wrappers/source/event.o \
+            system_wrappers/source/file_impl.o \
+            system_wrappers/source/list_no_stl.o \
+            system_wrappers/source/map.o \
+            system_wrappers/source/rw_lock.o \
+            system_wrappers/source/sort.o \
+            system_wrappers/source/thread.o \
+            system_wrappers/source/trace_impl.o
+
+NSX_OBJS := modules/audio_processing/ns/main/source/noise_suppression.o \
+	    modules/audio_processing/ns/main/source/ns_core.o
+
+export WEBRTCAEC_SRCDIR = $(WEBRTC_SRC_BASE)/src
+export WEBRTCAEC_OBJS = $(AEC_OBJS) $(NSX_OBJS) $(APM_OBJS) $(SWR_OBJS) $(SPL_OBJS)
+export WEBRTCAEC_CFLAGS = $(_CFLAGS)
+export WEBRTCAEC_CXXFLAGS = $(_CXXFLAGS)
+
+export CC_OUT CC AR RANLIB HOST_MV HOST_RM HOST_RMDIR HOST_MKDIR OBJEXT LD LDOUT
+
+
+###############################################################################
+# Main entry
+#
+# $(TARGET) is defined in os-$(OS_NAME).mak file in current directory.
+#
+TARGETS := libwebrtcaec
+
+all: $(TARGETS)
+
+doc:
+	cd .. && doxygen docs/doxygen.cfg
+
+dep: depend
+distclean: realclean
+
+.PHONY: dep depend libwebrtcaec clean realclean distclean
+
+libwebrtcaec:
+	$(MAKE) -f $(RULES_MAK) APP=WEBRTCAEC app=libwebrtcaec $(WEBRTCAEC_LIB)
+
+clean print_lib:
+	$(MAKE) -f $(RULES_MAK) APP=WEBRTCAEC app=libwebrtcaec $@
+
+realclean:
+	$(subst @@,$(subst /,$(HOST_PSEP),.ilbc-$(TARGET_NAME).depend),$(HOST_RMR))
+	$(MAKE) -f $(RULES_MAK) APP=WEBRTCAEC app=libwebrtcaec $@
+
+depend:
+	$(MAKE) -f $(RULES_MAK) APP=WEBRTCAEC app=libwebrtcaec $@
+
+
--- third_party/build/webrtc/os-auto.mak.in	1970-01-01 01:00:00.000000000 +0100
+++ third_party/build/webrtc/os-auto.mak.in	2011-09-13 11:56:08.000000000 +0200
@@ -0,0 +1,42 @@
+# @configure_input@
+
+AC_WEBRTC_PLATFORM=@ac_webrtc_platform@
+
+ifeq ($(AC_WEBRTC_PLATFORM),webrtc_linux)
+export CFLAGS += -DWEBRTC_TARGET_PC -DWEBRTC_LINUX -DWEBRTC_THREAD_RR
+export CFLAGS += -pthread
+
+export WEBRTCAEC_OBJS += system_wrappers/source/condition_variable_posix.o \
+            		 system_wrappers/source/critical_section_posix.o \
+            		 system_wrappers/source/event_posix.o \
+            		 system_wrappers/source/rw_lock_posix.o \
+            		 system_wrappers/source/thread_posix.o \
+            		 system_wrappers/source/trace_posix.o \
+            		 system_wrappers/source/cpu_linux.o
+endif
+
+ifeq ($(AC_WEBRTC_PLATFORM),webrtc_darwinos)
+export CFLAGS += -DWEBRTC_TARGET_MAC_INTEL -DWEBRTC_MAC_INTEL -DWEBRTC_MAC -DWEBRTC_THREAD_RR -DWEBRTC_CLOCK_TYPE_REALTIME
+export CFLAGS += -pthread
+
+export WEBRTCAEC_OBJS += system_wrappers/source/condition_variable_posix.o \
+            		 system_wrappers/source/critical_section_posix.o \
+            		 system_wrappers/source/event_posix.o \
+            		 system_wrappers/source/rw_lock_posix.o \
+            		 system_wrappers/source/thread_posix.o \
+            		 system_wrappers/source/trace_posix.o \
+            		 system_wrappers/source/cpu_mac.o
+endif
+
+ifeq ($(AC_WEBRTC_PLATFORM),webrtc_win32)
+export CFLAGS += -DWEBRTC_TARGET_PC -D_WIN32 -D_CRT_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_DEPRECATE -D__STD_C
+
+export WEBRTCAEC_OBJS += system_wrappers/source/condition_variable_windows.o \
+            		 system_wrappers/source/critical_section_windows.o \
+            		 system_wrappers/source/event_windows.o \
+            		 system_wrappers/source/rw_lock_windows.o \
+            		 system_wrappers/source/thread_windows.o \
+            		 system_wrappers/source/trace_windows.o \
+            		 system_wrappers/source/cpu_windows.o
+endif
+
--- third_party/build/webrtc/os-darwinos.mak	1970-01-01 01:00:00.000000000 +0100
+++ third_party/build/webrtc/os-darwinos.mak	2011-09-13 11:16:25.000000000 +0200
@@ -0,0 +1,12 @@
+
+export CFLAGS += -DWEBRTC_TARGET_MAC_INTEL -DWEBRTC_MAC_INTEL -DWEBRTC_MAC -DWEBRTC_THREAD_RR -DWEBRTC_CLOCK_TYPE_REALTIME
+export CFLAGS += -pthread
+
+export WEBRTCAEC_OBJS += system_wrappers/source/condition_variable_posix.o \
+            		 system_wrappers/source/critical_section_posix.o \
+            		 system_wrappers/source/event_posix.o \
+            		 system_wrappers/source/rw_lock_posix.o \
+            		 system_wrappers/source/thread_posix.o \
+            		 system_wrappers/source/trace_posix.o \
+            		 system_wrappers/source/cpu_mac.o
+
--- third_party/build/webrtc/os-linux.mak	1970-01-01 01:00:00.000000000 +0100
+++ third_party/build/webrtc/os-linux.mak	2011-09-13 11:55:35.000000000 +0200
@@ -0,0 +1,12 @@
+
+export CFLAGS += -DWEBRTC_TARGET_PC -DWEBRTC_LINUX -DWEBRTC_THREAD_RR
+export CFLAGS += -pthread
+
+export WEBRTCAEC_OBJS += system_wrappers/source/condition_variable_posix.o \
+            		 system_wrappers/source/critical_section_posix.o \
+            		 system_wrappers/source/event_posix.o \
+            		 system_wrappers/source/rw_lock_posix.o \
+            		 system_wrappers/source/thread_posix.o \
+            		 system_wrappers/source/trace_posix.o \
+            		 system_wrappers/source/cpu_linux.o
+
--- third_party/build/webrtc/os-win32.mak	1970-01-01 01:00:00.000000000 +0100
+++ third_party/build/webrtc/os-win32.mak	2011-09-13 11:01:32.000000000 +0200
@@ -0,0 +1,11 @@
+
+export CFLAGS += -DWEBRTC_TARGET_PC -D_WIN32 -D_CRT_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_DEPRECATE -D__STD_C
+
+export WEBRTCAEC_OBJS += system_wrappers/source/condition_variable_windows.o \
+            		 system_wrappers/source/critical_section_windows.o \
+            		 system_wrappers/source/event_windows.o \
+            		 system_wrappers/source/rw_lock_windows.o \
+            		 system_wrappers/source/thread_windows.o \
+            		 system_wrappers/source/trace_windows.o \
+            		 system_wrappers/source/cpu_windows.o
+
