--- pjsip/src/pjsip-ua/sip_inv.c	2009-04-24 12:29:40.000000000 +0200
+++ pjsip/src/pjsip-ua/sip_inv.c.fix	2009-04-23 18:23:31.000000000 +0200
@@ -3727,7 +3727,8 @@
 	    /* Handle response that terminates dialog */
 	    /* Nothing to do (already handled) */
 
-	} else if (tsx->status_code >= 300 && tsx->status_code < 700) {
+	} else if (tsx->state == PJSIP_TSX_STATE_COMPLETED &&
+			tsx->status_code >= 300 && tsx->status_code < 700) {
 
 	    pjmedia_sdp_neg_state neg_state;
 
@@ -3736,6 +3737,8 @@
 	    if (neg_state == PJMEDIA_SDP_NEG_STATE_LOCAL_OFFER) {
 		pjmedia_sdp_neg_cancel_offer(inv->neg);
 	    }
+
+	    inv->invite_tsx = NULL;
 	}
 
     } else if (tsx->role == PJSIP_ROLE_UAS &&
