diff -uNr pjsip_svn_export/pjsip/src/pjsip/sip_transport_tls_ossl.c pjsip_svn_export_work/pjsip/src/pjsip/sip_transport_tls_ossl.c
--- pjsip/src/pjsip/sip_transport_tls_ossl.c	2008-12-23 18:27:53.000000000 +0100
+++ pjsip/src/pjsip/sip_transport_tls_ossl.c	2011-05-27 16:28:10.000000000 +0200
@@ -345,18 +345,9 @@
 	method = PJSIP_SSL_DEFAULT_METHOD;
 
     switch (method) {
-    case PJSIP_SSLV23_METHOD:
-	ssl_method = SSLv23_method();
-	break;
     case PJSIP_TLSV1_METHOD:
 	ssl_method = TLSv1_method();
 	break;
-    case PJSIP_SSLV2_METHOD:
-	ssl_method = SSLv2_method();
-	break;
-    case PJSIP_SSLV3_METHOD:
-	ssl_method = SSLv3_method();
-	break;
     default:
 	ssl_report_error(lis_name, 4, PJSIP_TLS_EINVMETHOD,
 			 "Error creating SSL context");
