--- pjlib/src/pj/os_core_unix.c	2008-12-24 10:17:08.000000000 +0100
+++ pjlib/src/pj/os_core_unix.c	2011-09-13 09:58:04.000000000 +0200
@@ -1231,7 +1231,10 @@
     PJ_ASSERT_RETURN(mutex, PJ_EINVAL);
 
 #if PJ_DEBUG
-    pj_assert(mutex->owner == pj_thread_this());
+    //pj_assert(mutex->owner == pj_thread_this());
+    if (mutex->owner != pj_thread_this()) {
+        PJ_LOG(5,(mutex->obj_name, "Mutex released by another thread! %s", pj_thread_this()->obj_name));
+    }
     if (--mutex->nesting_level == 0) {
 	mutex->owner = NULL;
 	mutex->owner_name[0] = '\0';

