--- pjmedia/src/pjmedia/sdp_neg.c	2012-05-21 09:44:57.000000000 +0200
+++ pjmedia/src/pjmedia/sdp_neg.c	2013-01-07 14:00:46.000000000 +0100
@@ -783,7 +783,7 @@
 
 
     /* Process direction attributes */
-    update_media_direction(pool, answer, offer);
+    /*update_media_direction(pool, answer, offer);*/
  
     /* If asymetric media is allowed, then just check that remote answer has 
      * codecs that are within the offer. 

