diff -r 72f180130a42 build/temp.win32-2.6/Release/pjsip/third_party/build/portaudio/os-auto.mak.in
--- third_party/build/portaudio/os-auto.mak.in	Wed Aug 11 19:38:40 2010 +0200
+++ third_party/build/portaudio/os-auto.mak.in	Sat Aug 14 11:42:35 2010 +0200
@@ -75,5 +75,5 @@
 ifeq ($(AC_PJMEDIA_SND),pa_win32)
 export PORTAUDIO_OBJS += pa_win_hostapis.o pa_win_util.o \
 		       pa_win_wmme.o pa_win_waveformat.o
-export CFLAGS += -DPA_NO_ASIO -DPA_NO_DS
+export CFLAGS += -DPA_NO_ASIO -DPA_NO_DS -mwindows -static-libstdc++
 endif
