--- pjsip/src/pjsip-ua/sip_inv.c	2013-02-18 14:11:52.000000000 +0100
+++ pjsip/src/pjsip-ua/sip_inv.c	2013-02-18 14:13:16.000000000 +0100
@@ -1727,6 +1727,15 @@
 	}
     }
 
+    /* Cancel SDP negotiation if this is a negative reply to a re-INVITE */
+    if (st_code >= 300 && inv->neg != NULL &&
+            inv->state == PJSIP_INV_STATE_CONFIRMED) {
+        pjmedia_sdp_neg_state neg_state;
+        neg_state = pjmedia_sdp_neg_get_state(inv->neg);
+        if (neg_state == PJMEDIA_SDP_NEG_STATE_REMOTE_OFFER) {
+            pjmedia_sdp_neg_cancel_offer(inv->neg);
+        }
+    }
 
     return PJ_SUCCESS;
 }
