Index: pjmedia/src/pjmedia-codec/g722/g722_enc.c
===================================================================
--- pjmedia/src/pjmedia-codec/g722/g722_enc.c	(revision 2990)
+++ pjmedia/src/pjmedia-codec/g722/g722_enc.c	(working copy)
@@ -543,7 +543,7 @@
     PJ_ASSERT_RETURN(*out_size >= (nsamples >> 1), PJ_ETOOSMALL);
     
     for(i = 0; i < nsamples; i += 2) {
-	tx_qmf(enc, in[i]>>2, in[i+1]>>2, &xlow, &xhigh);
+	tx_qmf(enc, in[i], in[i+1], &xlow, &xhigh);
 
 	/* low band encoder */
 	ilow = block1l (xlow, enc->slow, enc->detlow) ;
Index: pjmedia/src/pjmedia-codec/g722/g722_dec.c
===================================================================
--- pjmedia/src/pjmedia-codec/g722/g722_dec.c	(revision 2990)
+++ pjmedia/src/pjmedia-codec/g722/g722_dec.c	(working copy)
@@ -530,8 +530,8 @@
 	/* rhigh <= output high band pcm */
 
 	rx_qmf(dec, rlow, rhigh, &pcm1, &pcm2);
-	out[i*2] = (pj_int16_t)(pcm1 << 2);
-	out[i*2+1] = (pj_int16_t)(pcm2 << 2);
+	out[i*2] = (pj_int16_t)pcm1;
+	out[i*2+1] = (pj_int16_t)pcm2;
     }
 
     *nsamples = in_size << 1;
